shared/markdown/parser.pegjs

```diff
@@ -1,154 +1,161 @@
 {
-  const linkExp = new RegExp(/(?:(?:ftp|http(?:s)?)?:\/\/.)?(?:www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b(?:[-a-zA-Z0-9@:%_\+.~#?&\/\/=]*)\b/i)
-  const protoExp = new RegExp(/^([a-z]*):\/\//i)
-  const dotDotExp = new RegExp(/[^/]\.\.[^/]/)
-  const mailToExp = new RegExp(/^mailto:/i)
   // Instead of encoding all the bad cases into a more complicated regexp lets just add some simple code here
   // Note: We aren't trying to be 100% perfect here, just getting something that works pretty good and pretty quickly
-  function goodLink (link, protocolMatch) {
-    return !link.match(dotDotExp) && // disallow 'a...b', but allow /../
-      !link.match(mailToExp) && // disallow mailto:
-      (!protocolMatch || ['http://', 'https://'].includes(protocolMatch[0].toLowerCase())) // only allow http(s)
+  function goodLink (link) {
+    return !link.match(dotDotExp) // disallow 'a...b', but allow /../
   }
 
-  function convertLink (text) {
-    const matches = text.match(linkExp)
-    if (matches) {
-      const match = matches[0]
-      const protocolMatch = match.match(protoExp)
-      if (goodLink(match, protocolMatch)) {
-        const href = protocolMatch && match || 'http://' + match
-        const start = matches.index
-        const end = start + match.length
-        const left = text.substring(0, start)
-        const right = text.substring(start + end)
-        return {
-          type: 'text',
-          children: [
-            ...(left ? [left] : []),
-            {type: 'link', children: [match], href},
-            ...(right ? [right] : []),
-          ],
+  function flatten (input) {
+    const result = []
+    let strs = []
+
+    function visit(x) {
+      if (Array.isArray(x) ) {
+        for (const y of x) {
+          if (y) {
+            visit(y)
+          }
+        }
+      } else if (typeof x === 'string') {
+        strs.push(x)
+      } else {
+        if (strs.length) {
+          result.push(strs.join(''))
+          strs = []
         }
+        result.push(x)
       }
     }
-    return text
+
+    visit(input)
+    if (strs.length) {
+      result.push(strs.join(''))
+    }
+    return result
   }
 }
 
 start
- = children:(Blank / Code / Content / Blank)* { return {type: 'text', children}; }
+ = children:(QuoteBlock / Line / BlankLine)* { return {type: 'text', children: flatten(children)} }
 
-Code = CodeBlock / InlineCode
+BlankLine
+ = WhiteSpace* LineTerminatorSequence { return text() }
 
-Content
- = StyledText / Text
+Line
+ = (WhiteSpace* __INLINE_MACRO__<> / WhiteSpace+) LineTerminatorSequence?
```

Comment:
The primary change in the parser is using the syntax "special characters" to delimit text blocks. Previously, when parsing `hello*world*`, the `Text` rule would consume all of the characters. Now it consumes `hello` and another rule can match `*world*`.

This lets us generalize the rules, using a single main `Inline` rule which recurses, instead of `Bold / FromBold / InsideBoldMarker`. For comparison:

previous (simplified):
```
Text = NonBlank+
Bold = BoldMarker children:FromBold* BoldMarker
FromBold = (Italic / Strike / InsideBoldMarker)
InsideBoldMarker = (! BoldMarker .)
```

next (simplified):
```
Text = (!NativeEmojiCharacter !SpecialChar NonBlank)+
BoldMarker (!BoldMarker Inline)+ !BoldMarker
```

This works because `Text` now stops at every special character like `*`. So repeating a rule like `(!BoldMarker Inline)+`, we are assured to stop at the next `*`. 

However, there is one hitch: a nice feature of the previous approach is that it was space-delimited. `Text` would consume all of `hello*world*`, precluding the possibility of it being matched by the `Bold` rule.

To regain this behavior, we need to split the `Inline` rule, separating out things that must start after a whitespace character. `InlineStart` contains space-delimited rules such as italic, bold, and links. `InlineCont` contains rules which can consume characters without a space in between. Finally, we must rewrite `(!BoldMarker Inline)+` to be a bit more complex: `!BoldMarker InlineStart ((WhiteSpace+ !BoldMarker InlineStart) / (!BoldMarker InlineCont))`. The `__INLINE_MACRO__<delimiter>` outputs this rule for a given delimiter character.